File: src/eqn/Eqn.java
package eqn;

import eqn.ast.EqnAst;
import eqn.parser.exception.EqnException;
import org.jetbrains.annotations.NotNull;

public class Eqn {

    final private EqnHead equationHead;
    final private EqnAst equationAst;

    public Eqn(@NotNull EqnHead equationHead, @NotNull EqnAst equationAst) throws EqnException {
        this.equationHead = equationHead;
        this.equationAst = equationAst;
        this.equationAst.simplify();
        System.out.println("Equation created:");
        System.out.println("\t" + equationHead.toString());
        System.out.println("\t" + equationAst.toString());
    }

    public Double evaluate() throws EqnException {
        return equationAst.evaluate();
    }

}

File: src/eqn/Eqn.kt
package eqn

import eqn.ast.base.EqnAst
import eqn.parser.exception.EqnException

class Eqn(private val equationHead: EqnHead, private val equationAst: EqnAst) {
    override fun toString(): String {
        return "Equation created:\n\t$equationHead\n\t$equationAst"
    }

    @Throws(EqnException::class)
    fun evaluate(arguments: Map<String, Double>?): Double {
        val variablePredicate: (String) -> Boolean = { equationHead.getVariableNames().contains(it) }
        val filteredArguments = arguments!!.filterKeys(variablePredicate)
        if (filteredArguments.size != equationHead.getVariableNames().size) {
            throw IllegalArgumentException("Trying to evaluate in Eqn without setting all variables!")
        }
        return equationAst.evaluate(filteredArguments)
    }

    fun simplify(arguments: Map<String, Double>?) {
        val variablePredicate: (String) -> Boolean = { equationHead.getVariableNames().contains(it) }
        val filteredArguments = arguments!!.filterKeys(variablePredicate)
        equationAst.simplify(filteredArguments)
    }

    init {
        equationAst.simplify(null)
        println("Equation created:\n\t$equationHead\n\t$equationAst")
    }
}